pkg/controller/node/node_controller.go

```diff
@@ -0,0 +1,140 @@
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("node-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	labelPreds := []predicate.Predicate{
+		predicate.Funcs{
+			UpdateFunc: func(evt event.UpdateEvent) bool {
+				// get node with the update event name
+				rNode := r.(*ReconcileNode)
+				instance := &corev1.Node{}
+				rNode.client.Get(context.Background(), types.NamespacedName{Name: evt.MetaOld.GetName()}, instance)
+				watchResourceTagLabel, watchOk := instance.GetLabels()[watchResourceTag]
+
+				return !(watchOk && watchResourceTagLabel == watchResourceValue)
+			},
+			GenericFunc: func(evt event.GenericEvent) bool {
+				rNode := r.(*ReconcileNode)
+				instance := &corev1.Node{}
+				rNode.client.Get(context.Background(), types.NamespacedName{Name: evt.Meta.GetName()}, instance)
+				watchResourceTagLabel, watchOk := instance.GetLabels()[watchResourceTag]
+
+				return !(watchOk && watchResourceTagLabel == watchResourceValue)
+			},
+		},
+	}
+
+	// Watch for changes to primary resource Node
+
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}, labelPreds...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNode implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNode{}
+
+// ReconcileNode reconciles a Node object
+type ReconcileNode struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Node object and makes changes based on the state read
+// and what is in the Node.Spec
+func (r *ReconcileNode) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+	reqLogger.Info("Reconciling Node")
+
+	// Fetch the Node instance
+	instance := &corev1.Node{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Error(err, "node does not exist")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "Failed to get node")
+		return reconcile.Result{}, err
+	}
+
+	labels := instance.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[watchResourceTag] = watchResourceValue
+	instance.SetLabels(labels)
+
+	if err := r.client.Update(context.TODO(), instance); err != nil {
```

Comment:
Please add an `if` condition to check if an update is required. Even if we filter via the watch, we should still have an `if` check here.

Reply:
I can add the condition but according to the watch only the nodes which doesn't have the label `razee/watch-resource ` gets updated right. So is there any chance where watch fails and it gets here even the label present ?

Reply:
There isn't a chance, but adding an if protects future devs who may change the watch conditions.

Reply:
Sounds good 👍🏻